### PR TITLE
[ALLUXIO-2088] Replace single-element IndexedSet with FieldIndex

### PR DESCRIPTION
**Pull request by xhad1234**
Some indexedSet only has one element. We can directly use an index instead of indexedSet for this case. Testing shows memory utilization improvement by removing those indexedSet.

For details, please see: https://alluxio.atlassian.net/browse/ALLUXIO-2088

### CODE REVIEW

**Comment by xhad1234** [comment]
alluxio-bot, check this please

File: core/server/src/main/java/alluxio/master/file/meta/InodeDirectory.java [hunk 1]
```diff
@@ -76,15 +77,20 @@ protected InodeDirectory getThis() {
    * @param child the inode to add
    */
   public void addChild(Inode<?> child) {
-    mChildren.add(child);
+    synchronized (child) {
```
Comment (gpang) [inline core/server/src/main/java/alluxio/master/file/meta/InodeDirectory.java]: I think this is one of the problems with using index directly. It makes users of the index much more complicated and error prone.

Comment (xhad1234) [inline core/server/src/main/java/alluxio/master/file/meta/InodeDirectory.java]: We can hide this synchronize if we let the index add returns false when facing a duplicated value instead of throwing an exception. Should we do that?

Comment (gpang) [inline core/server/src/main/java/alluxio/master/file/meta/InodeDirectory.java]: users should not need to worry about synchronization. Originally, IndexedSet handled synchronization across indexes.

If we allow users to also use index directly, we need to ensure that indexes are easy to use, and are thread-safe. What is the thread-safety of `FieldIndex`?

**Comment by gpang** [comment]
@xhad1234 Could you resolve the merge conflicts?

I also had a question about `FieldIndex`. Thanks!

File: core/server/src/main/java/alluxio/master/file/meta/InodeDirectory.java (hunk 1 above)
Comment (xhad1234) [inline core/server/src/main/java/alluxio/master/file/meta/InodeDirectory.java]: Gene, since there is only one concurrentHashMap inside each index, it is thread-safe without synchronized.

**Comment by xhad1234** [comment]
@gpang ,thank you for your comment. I hide the synchronization from user and also merge the conflicts.

File: core/common/src/main/java/alluxio/collections/FieldIndex.java [hunk 2]
```diff
@@ -22,13 +22,14 @@
  *
  * @param <T> type of objects in this {@link IndexedSet}
  */
-interface FieldIndex<T> extends Iterable<T> {
+public interface FieldIndex<T> extends Iterable<T> {
   /**
    * Adds an object o to the index.
    *
    * @param o the object to add to the index
+   * @return true if success, otherwise false
```
Comment (gpang) [inline core/common/src/main/java/alluxio/collections/FieldIndex.java]: Could you be a bit more explicit in what success means? For example Java's `Set.add`: https://docs.oracle.com/javase/7/docs/api/java/util/Set.html#add(E)

File: core/common/src/main/java/alluxio/collections/IndexedSet.java [hunk 3]
```diff
@@ -292,7 +294,6 @@ public boolean remove(Object object) {
     if (object == null) {
       return false;
     }
-    // Locking this object protects against removing the exact object that might be in the
```
Comment (gpang) [inline core/common/src/main/java/alluxio/collections/IndexedSet.java]: Why is this line removed?

File: core/common/src/main/java/alluxio/collections/NonUniqueFieldIndex.java [hunk 4]
```diff
@@ -62,6 +64,7 @@ public void add(T object) {
         break;
```
Comment (gpang) [inline core/common/src/main/java/alluxio/collections/NonUniqueFieldIndex.java]: On the previous line, would the `add` ever return false?

File: core/common/src/main/java/alluxio/collections/UniqueFieldIndex.java [hunk 5]
```diff
@@ -25,26 +25,34 @@
...
+   * @param indexDefinition definition of index
    */
   public UniqueFieldIndex(IndexDefinition<T> indexDefinition) {
     mIndexMap = new ConcurrentHashMap<>(8, 0.95f, 8);
     mIndexDefinition = indexDefinition;
   }
 
   @Override
-  public void add(T object) {
+  public boolean add(T object) {
```
Comment (gpang) [inline core/common/src/main/java/alluxio/collections/UniqueFieldIndex.java]: I think the behavior of this `add` is slightly different now. Is there already a unittest for `IndexedSet` that covers this add for this class?

File: core/common/src/main/java/alluxio/collections/UniqueFieldIndex.java [hunk 6]
```diff
@@ -25,26 +25,34 @@
...
     Object fieldValue = mIndexDefinition.getFieldValue(object);
-
-    if (mIndexMap.putIfAbsent(fieldValue, object) != null) {
-      throw new IllegalStateException("Adding more than one value to a unique index:"
-          + fieldValue.toString());
+    T previousObject = mIndexMap.putIfAbsent(fieldValue, object);
+
+    if (previousObject != null) {
```
Comment (gpang) [inline core/common/src/main/java/alluxio/collections/UniqueFieldIndex.java]: I think this block would be easier to read if it were something like:

```
if (previousObject == null) {
  return true;
}
if (previousObject == object) {
  // Adding the same object again.
  return false;
}
throw ...
```

**Comment by gpang** [comment]
@xhad1234 I left a few comments. Thanks!

Also, I had a question. `InodeDirectory` and `InodeTree` both use the field index now. Is it correct to assume that the `InodeDirectory` change affects the memory utilization more?

File: core/common/src/main/java/alluxio/collections/NonUniqueFieldIndex.java (hunk 4 above)
Comment (xhad1234) [inline core/common/src/main/java/alluxio/collections/NonUniqueFieldIndex.java]: for non-unique index, they will not return false

File: core/common/src/main/java/alluxio/collections/UniqueFieldIndex.java (hunk 6 above)
Comment (xhad1234) [inline core/common/src/main/java/alluxio/collections/UniqueFieldIndex.java]: Thank you for the advice, updated.

Comment (apc999) [inline core/common/src/main/java/alluxio/collections/UniqueFieldIndex.java]: why we want to return `false` if adding the same object again while throw exception if adding more than one value to a unique index?

File: core/server/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 7]
```diff
@@ -102,7 +103,7 @@ public Object getFieldValue(Inode<?> o) {
   private final MountTable mMountTable;
 
   @SuppressWarnings("unchecked")
-  private final IndexedSet<Inode<?>> mInodes = new IndexedSet<>(ID_INDEX);
+  private FieldIndex<Inode<?>> mInodes = new UniqueFieldIndex<Inode<?>>(ID_INDEX);
```
Comment (apc999) [inline core/server/src/main/java/alluxio/master/file/meta/InodeTree.java]: why remove `final`?

Comment (apc999) [inline core/server/src/main/java/alluxio/master/file/meta/InodeTree.java]: Also I suggest to add more comments here indicating why we use `UniqueFieldIndex` directly rather than `IndexedSet`

File: core/common/src/main/java/alluxio/collections/FieldIndex.java (hunk 2 above)
Comment (xhad1234) [inline core/common/src/main/java/alluxio/collections/FieldIndex.java]: @gpang thank you for the comment, finally I changed this function back as returning void. Because if we a remove may happen before we return true thus make things inconsistent.

File: core/common/src/main/java/alluxio/collections/IndexedSet.java (hunk 3 above)
Comment (xhad1234) [inline core/common/src/main/java/alluxio/collections/IndexedSet.java]: Added back

File: core/common/src/main/java/alluxio/collections/UniqueFieldIndex.java (hunk 5 above)
Comment (xhad1234) [inline core/common/src/main/java/alluxio/collections/UniqueFieldIndex.java]: @gpang ,thank you for the comments. I changed the add function back. Instead of returning false,  it would just return without doing anything when meeting two duplicated objects.

File: core/server/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 7 above)
Comment (xhad1234) [inline core/server/src/main/java/alluxio/master/file/meta/InodeTree.java]: @apc999 thank you for the comments. `final` and comments added

**Comment by xhad1234** [comment]
@gpang Thank you for the comments, I did some updates. As for the question, yes, `InodeDirectory`  affects memory utilization more.

**Comment by xhad1234** [comment]
@apc999 Thank you for the comments, I updated the code.

**Comment by apc999** [comment]
LGTM. thanks for the improvement

**Comment by gpang** [comment]
LGTM
